### PR TITLE
webgpu: add leakyrelu operator

### PR DESCRIPTION
**Pull request by justmywyw**
To see the logs from the Cloud Build CI, please join either our [discussion](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs) or [announcement](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs-announce) mailing list.

---
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/5933)

### CODE REVIEW

File: tfjs-backend-webgpu/src/kernels/LeakyRelu.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+ export function leakyRelu(args:
+  {inputs: LeakyReluInputs, backend: WebGPUBackend, attrs: LeakyReluAttrs}):
+ TensorInfo {
+   const {inputs, backend, attrs} = args;
+   const {x} = inputs;
+   const {alpha} = attrs;
+   const program = new LeakyReluProgram(x.shape);
+   const uniformData = [{type: 'float32', data: [alpha]}];
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/kernels/LeakyRelu.ts:68]: Nice. I like the change that passing alpha as a uniform instead a tensor. Can you reuse the `UnaryOpProgram`?
```
const program = new UnaryOpProgram(x.shape, UnaryOpType.LEAKYRELU);
program.uniforms = 'alpha : f32;'
```
The leakyRelu string will be like `if (a < 0.0) { return uniforms.alpha * a; } return a;`

File: tfjs-backend-webgpu/src/setup_test.ts [hunk 2]
```diff
@@ -115,6 +115,9 @@ const TEST_FILTERS: TestFilter[] = [
       'leakyrelu'           // Not yet implemented.
     ]
   },
+  {
+    include: 'leakyrelu',
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/setup_test.ts:119]: nit: maybe using `startsWith` is preferable. I remember fused depthwise/conv2d/matmul with `leakyrelu` is still unsupported for webgpu. And it's welcome for you to submit another PR to support `leakyrelu` to those ops and remove `leakyrelu` from the `excludes` in this file.

**Review by qjia7 — COMMENTED** [review]
Thanks for the contribution. @justmywyw

File: tfjs-backend-webgpu/src/kernels/LeakyRelu.ts (hunk 1 above)
Comment (justmywyw) [reply tfjs-backend-webgpu/src/kernels/LeakyRelu.ts]: Yes，and finish it on the last commit.

File: tfjs-backend-webgpu/src/setup_test.ts (hunk 2 above)
Comment (justmywyw) [reply tfjs-backend-webgpu/src/setup_test.ts]: FIxed here. And I`m glad to work on to support leakyrelu to those other ops ~

File: tfjs-backend-webgpu/src/setup_test.ts [hunk 3]
```diff
@@ -115,6 +115,9 @@ const TEST_FILTERS: TestFilter[] = [
       'leakyrelu'           // Not yet implemented.
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/setup_test.ts:115]: Please remove `leakyrelu` from `excludes` of `relu`. And change `include: 'relu'` -> `startsWith: 'relu'`. Otherwise, `leakyrelu` is not really tested. Similarly, change `include: 'elu'` -> `startsWith: 'elu'`. There may be some other failures due to this change. You can just exclude those cases in the corresponding category and add the failure reason.

**Review by qjia7 — APPROVED** [review]
LGTM for the last change. Thanks.

File: tfjs-backend-webgpu/src/setup_test.ts (hunk 3 above)
Comment (justmywyw) [reply tfjs-backend-webgpu/src/setup_test.ts]: Checked all the test case of leakyRelu  and exclude one failure case.

However  some other failure blocked at the beginning, cause `dot(vecN<i32>, vecN<i32>)` (https://github.com/tensorflow/tfjs/blob/master/tfjs-backend-webgpu/src/shader_preprocessor.ts#L272)  not supported ( find it recently added [here](https://github.com/gpuweb/gpuweb/issues/2231)) and it lead to wgsl parse error, maybe it's not supported for Metal.
And I temporarily remove them and successfully executed.

Comment (qjia7) [reply tfjs-backend-webgpu/src/setup_test.ts]: I still meet some errors for `fused` operators with `gradient` tests. You need to add `'gradient',   // Kernel 'Step' not registered`  in `excludes` part of `include: 'fused',`. Sorry that it's a mess to  use `includes` at the beginning. We properly need a refactor for the test filtering later.

Are you using the latest chrome canary in Mac? If `dot(vecN<i32>, vecN<i32>)` issue still exists in latest chrome, we need to report a bug to tint to support.

Comment (justmywyw) [reply tfjs-backend-webgpu/src/setup_test.ts]: Sorry that the dot(vecN<i32>, vecN<i32>) issue exists in chrome 96.
(Additionally, Chrome has started webgpu trial in Chrome 94，wondering weather the compatibility should be considered.)

And I've tried chrome 98 canary and update it to the latest just now and try `yarn test` again, but there still many other failure block the test, seems like fail error aborted sometimes and lead to webgpu backend initialize fail, maybe I missing something.

As for the errors for `fused` operators with `gradient` tests，wondering if they are associated with the leakyRelu ops, cause the tests cases for `fused` who are including `leakyrelu` executed successfully in my env.

Comment (qjia7) [reply tfjs-backend-webgpu/src/setup_test.ts]: @justmywyw I haven't tried Mac's `yarn test` situation previously. But on windows, we have 100% pass rate with current filter rule. I will meet below error after applying your PR:
```
FAILED fused matmul webgpu {"WEBGPU_CPU_FORWARD":false} fused A x B with relu bias gradient
debug.js:21 Error: Kernel 'Step' not registered for backend 'webgpu'
FAILED fused matmul webgpu {"WEBGPU_CPU_FORWARD":false} gradient with clones A x B with relu
Error: Kernel 'Step' not registered for backend 'webgpu'
FAILED fused matmul webgpu {"WEBGPU_CPU_FORWARD":false} fused matmul with relu6 and gradients
Error: Kernel 'Step' not registered for backend 'webgpu'
FAILED fused matmul webgpu {"WEBGPU_CPU_FORWARD":false} fused A x B with relu and broadcasted bias gradient
Error: Kernel 'Step' not registered for backend 'webgpu'
FAILED fused matmul webgpu {"WEBGPU_CPU_FORWARD":false} fused A x B with relu gradient
Error: Kernel 'Step' not registered for backend 'webgpu'
```
I think the error is caused by changing `include: 'relu'` to `startsWith: 'relu'`. So please add `'gradient', // Kernel 'Step' not registered` in `excludes` part of `include: 'fused'` so that all cases can pass on windows. After that, I will merge your PR, thanks.

I will check the situations on mac. Thanks for your report for the mac issue.

Comment (qjia7) [reply tfjs-backend-webgpu/src/setup_test.ts]: >there still many other failure block the test, seems like fatal error aborted sometimes and lead to webgpu backend initialize fail, maybe I missing something.

fyi, I just tried the latest chrome canary 99.0.4764.2 on Mac. I didn't meet the error that webgpu backend initialization fails. There are totally 3 failed cases (2 resizeNearestNeighbor cases and one image.transform case) in master branch. You can ignore these 3 cases for now.

Comment (justmywyw) [reply tfjs-backend-webgpu/src/setup_test.ts]: @qjia7 I add the exclude of `gradient` as your test result said.

However, I tried the latest chrome canary 99.0.4764.2 on my Mac, run `yarn test` on a new cloned tfjs project's master branch and receive the error below : 
````
Chrome 99.0.4764.2 (Mac OS 10.15.7): Executed 32 of 4128 (20 FAILED) (0 secs / 0Chrome 99.0.4764.2 (Mac OS 10.15.7) less webgpu {"WEBGPU_CPU_FORWARD":false} broadcasting Tensor4D shapes - float32 FAILED
	AbortError: Device is lost

Chrome 99.0.4764.2 (Mac OS 10.15.7) fused depthwiseConv2D webgpu {"WEBGPU_CPU_FORWARD":false} throws when input is int32 FAILED
	Error: The highest priority backend 'webgpu' has not yet been initialized. Make sure to await tf.ready() or await tf.setBackend() before calling other methods

````
It must be something wrong with my environment, and I‘ll try to figure out later and post here.

Comment (qjia7) [reply tfjs-backend-webgpu/src/setup_test.ts]: Are you running chrome canary with flag `--enable-unsafe-webgpu`?

**Review by qjia7 — APPROVED** [review]
LGTM, thanks.

cc @lina128 @pyu10055

File: tfjs-backend-webgpu/src/setup_test.ts (hunk 3 above)
Comment (justmywyw) [reply tfjs-backend-webgpu/src/setup_test.ts]: The flag `--enable-unsafe-webgpu` is added by karma, and I've checked the console that `GPU` exists on window.
And the test case fail after some success executed ( which means webgpu have been  initialized at the beginning), maybe some fatal error lead to gpu device lost.

Comment (qjia7) [reply tfjs-backend-webgpu/src/setup_test.ts]: Ah, you are right. The flag --enable-unsafe-webgpu is added by karma. I don't know the device lost reason. Maybe try another device to see if you can reproduce it. Please file a bug to us and attach your hardware information to see if we can reproduce it in our side.
